README

		$Id$

xs-nntp-slb is a simplistic NNTP loadbalancer.

It is mainly meant to be used as a frontend to a cluster of
NNTP transit servers, to spread out the load.

It can only do static loadbalancing; if a backend server goes
down, xs-nntp-slb stops working.

xs-nntp-slb consists of 2 parts: a daemon, written in C, that accepts
connections on a TCP port (usually 119, ofcourse) which spawns
and executes a worker process written in Go.

The worker process makes multiple outgoing connections,
one to each backend. xs-nntp-slb-go uses the XCLIENT command to
forward the IP address of the client to each backend.

At each supported NNTP command that takes a message-id as
an argument, the message-id is hashed to a 32-bits number N.
Then the command is forwarded to backend-server
number (N modulo number_of_servers) .



debian/changelog
xs-nntp-slb-go (1.3) jessie stretch; urgency=low

  * jessie and stretch

 -- Miquel van Smoorenburg (XS4ALL) <mikevs@xs4all.net>  Thu, 19 Apr 2018 16:19:45 +0200

xs-nntp-slb-go (1.2) wheezy jessie; urgency=low

  * make sure init script is included.

 -- Miquel van Smoorenburg (XS4ALL) <mikevs@xs4all.net>  Mon, 17 Aug 2015 19:08:27 +0200

xs-nntp-slb-go (1.1) wheezy jessie; urgency=low

  * remove logrotate stuff
  * add example files for logrotate and default

 -- Miquel van Smoorenburg (XS4ALL) <mikevs@xs4all.net>  Mon, 17 Aug 2015 18:47:56 +0200

xs-nntp-slb-go (1.0) wheezy jessie; urgency=low

  * rewrite in Go.
  * initial release.

 -- Miquel van Smoorenburg (XS4ALL) <mikevs@xs4all.net>  Wed, 18 Feb 2015 09:59:19 +0100


log.go
//
//	Package that provides an alternative to 'log'
//	The interface is similar (but not the same)
//	Can write to stderr or syslog or both
//
//	TODO: somehow log stack trace from panic() as well..
//
package main

import (
	"fmt"
	"log"
	"log/syslog"
	"os"
	"path/filepath"
	"time"
)

const (
	LogSyslog = 1 << iota
	LogStderr
)

const (
	_ = iota
	LogDebug
	LogInfo
	LogNotice
	LogError
)

type Slog struct {
	s *syslog.Writer
	t int
}

var Log Slog
var logLabel string
var logDebug bool

func (l *Slog) initSyslog() {
	s, err := syslog.New(syslog.LOG_NEWS|syslog.LOG_NOTICE, logLabel)
	if err != nil {
		panic("syslog.New failed")
	}
	l.s = s
	return
}

func (l *Slog) syslogWriteString(level int, msg string) {
	switch level {
		case LogDebug:  _ = l.s.Debug(msg)
		case LogInfo:   _ = l.s.Info(msg)
		case LogNotice: _ = l.s.Notice(msg)
		case LogError:  _ = l.s.Err(msg)
		default:        _ = l.s.Err(msg)
	}
}

func (l *Slog) stderrWriteString(level int, msg string) {
	t := time.Now()
	fmt.Fprintf(os.Stderr, "%s %s: %s\n",
		t.Format(time.Stamp), logLabel, msg)
}

func (l *Slog) WriteString(level int, msg string) {
	if (l.t & LogSyslog) != 0 {
		l.syslogWriteString(level, msg)
	}
	if (l.t & LogStderr) != 0 {
		l.stderrWriteString(level, msg)
	}
}

func (l *Slog) Write(msg []byte) (n int, err error) {
	l.WriteString(LogError, string(msg))
	n = len(msg)
	return
}

func (l *Slog) SetOutput(fl int) {
	switch {
		case (fl & LogSyslog) != 0:
		case (fl & LogStderr) != 0:
	}
	l.t = fl
}

func (l *Slog) Printf(level int, format string, a ...interface{}) {
	var s string
	if len(a) == 0 {
		s = format
	} else {
		s = fmt.Sprintf(format, a...)
	}
	l.WriteString(level, ChompString(s))
}

func (l *Slog) Debug(fmt string, a ...interface{}) {
	if (logDebug) {
		l.Printf(LogDebug, fmt, a...)
	}
}

func (l *Slog) Info(fmt string, a ...interface{}) {
	l.Printf(LogInfo, fmt, a...)
}

func (l *Slog) Notice(fmt string, a ...interface{}) {
	l.Printf(LogNotice, fmt, a...)
}

func (l *Slog) Error(fmt string, a ...interface{}) {
	l.Printf(LogError, fmt, a...)
}

func (l *Slog) Fatal(fmt string, a ...interface{}) {
	l.Printf(LogError, fmt, a...)
	os.Exit(1)
}

func init() {
	logLabel = filepath.Base(os.Args[0])
	Log.t = LogSyslog

	Log.initSyslog()

	log.SetOutput(&Log)
	log.SetFlags(0)
	log.SetPrefix("")
}



nntpqueue.go
package main

import (
	"strings"
	"sync/atomic"
	"sync"
)

type NNTPReq struct {
	line	 string
	cmd	 string
	msgid    string
	code     int
	ready	 bool
}

type NNTPQueue struct {
	queue     []*NNTPReq
	qlock     sync.Mutex
	wlock     sync.Mutex
	sess      *NNTPSession
	lastcode  int32
}

//
//	For certain commands - right now only for STAT - we
//	add the message-id to the reply. This is for testing.
//
func (req *NNTPReq) addmsgid() {
	if req.code == 430 && len(req.msgid) > 0 &&
	   strings.IndexByte(req.line, '<') < 0 {
		req.line = ChompString(req.line) + " " + req.msgid + "\r\n"
	}
}

//
//	Add a request to a queue. Possibly runs the queue.
//
func (q *NNTPQueue) Add(req *NNTPReq, run_queue bool) {
	q.qlock.Lock()
	defer q.qlock.Unlock()

	// Append this entry to the queue.
	q.queue = append(q.queue, req)
	if len(q.queue) == 1 && run_queue {
		q.run()
	}
}

//
//	Pop oldest entry off the queue and return it.
//
func (q *NNTPQueue) PopFirst() (r *NNTPReq) {
	q.qlock.Lock()
	defer q.qlock.Unlock()
	len := len(q.queue)
	if len > 0 {
		r = q.queue[0]
		q.queue = q.queue[1:len]
	}
	return
}

//
//	Examine the oldest entry. If it is ready, pop it
//	off the queue and 'run' it. Rinse and repeat.
//	Assume that q.qlock is already locked.
//
func (q *NNTPQueue) run() {

	if len(q.queue) == 0 || !q.queue[0].ready {
		return
	}
	q.wlock.Lock()

	Log.Debug("running queue, len is %d", len(q.queue))

	olen := len(q.queue)

	for {
		req := q.queue[0]
		q.queue = q.queue[1:]
		q.qlock.Unlock()

		// Do not copy replies to QUIT
		var err error
		if req.cmd != "quit" {
			req.addmsgid()
			err = q.sess.Write(req.line)
		}

		if err != nil {
			// whoops, remote client has gone away
			Log.Fatal("%s: lost connection(write, qlen=%d->%d): " +
				"%s on %s (FATAL)", q.sess.name, olen, q.Len(),
				err, ChompString(req.line))
		}

		q.qlock.Lock()
		if len(q.queue) == 0 || !q.queue[0].ready {
			atomic.StoreInt32(&q.lastcode, int32(req.code))
			break;
		}
	}

	err := q.sess.Flush()
	q.wlock.Unlock()
	if err != nil {
		// whoops, remote client has gone away
		Log.Fatal("%s: lost connection(flush, qlen=%d->%d): %s (FATAL)",
			q.sess.name, olen, len(q.queue), err)
	}
}

func (q *NNTPQueue) LastCode() int {
	return int(atomic.LoadInt32(&q.lastcode))
}

func (q *NNTPQueue) Run() {
	q.qlock.Lock()
	defer q.qlock.Unlock()
	q.run()
}

func (q *NNTPQueue) Len() int {
	q.qlock.Lock()
	defer q.qlock.Unlock()
	return len(q.queue)
}

//
//	Set entry to 'ready' and try to run the queue
//
func (q *NNTPQueue) Ready(r *NNTPReq) {
	q.qlock.Lock()
	defer q.qlock.Unlock()
	r.ready = true
	q.run()
}



xs-nntp-slb.c
// +build ignore

/*
 * xs-nntp-slb	NNTP listener- forks off xs-nntp-slb-go command
 *		for each incoming connection.
 *
 */

#define _XOPEN_SOURCE 600
#define _GNU_SOURCE

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <sys/poll.h>
#include <netdb.h>
#include <syslog.h>
#include <stdarg.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
#include <string.h>

#define MAXLISTEN	16

int do_syslog = 1;
int do_stderr = 1;
int do_debug = 0;

char *progname;

void do_log(int dbg, char *fmt, ...);

#define notice(fmt, x...) \
	do_log(0, fmt, ##x)

#define perrordie(fmt, x...) \
	do { \
		do_log(1, fmt, ##x); \
		exit(1); \
	} while(0)

#define debug(fmt, x...) \
	do_log(2, fmt, ##x)

void do_log(int mode, char *fmt, ...)
{
	char		buf[1024];
	char		*p;
	va_list		ap;

	if (mode == 2 && !do_debug)
		return;

	va_start(ap, fmt);
	vsnprintf(buf, sizeof(buf), fmt, ap);
	va_end(ap);

	p = strstr(buf, "\r\n");
	if (p) *p = 0;

	if (mode == 2) {
		fprintf(stderr, "DBG: %s\n", buf);
		return;
	}

	if (do_syslog)
		syslog(LOG_NOTICE, "%s", buf);

	if (do_stderr) {
		if (mode == 1)
			fprintf(stderr, "%s: ", progname);
		fprintf(stderr, "%s\n", buf);
	}
}

int do_poll(int fd, int events, int timeout)
{
	struct pollfd	pfd;
	int		r;

	pfd.fd = fd;
	pfd.events = events;
	r = poll(&pfd, 1, timeout * 1000);
	if (r == 0) {
		r = -1;
		errno = ETIMEDOUT;
	}

	return r;
}

int blocking(int fd, int block)
{
	int fl = fcntl(fd, F_GETFL);
	if (fl < 0)
		return -1;
	if (block)
		fl &= ~(O_NONBLOCK);
	else
		fl |= O_NONBLOCK;
	return fcntl(fd, F_SETFL, fl);
}

void parse_host_port(char *data, int passive,
			char **hostp, char **portp, char *buf, int bufsz)
{
	char *host = NULL;
	char *port = NULL;
	char *colon = NULL;
	char *h;

	*hostp = NULL;
	*portp = NULL;

	strncpy(buf, data, bufsz);
	buf[bufsz - 1] = 0;
	data = buf;

	if (*data == '[' && (h = strchr(data, ']')) != NULL) {
		data++;
		*h = 0;
		colon = (h[1] == ':') ? h + 1 : NULL;
	} else
		colon = strrchr(data, ':');

	if (colon) {
		host = data;
		*colon++ = 0;
		port = colon;
	} else {
		int i;
		for (i = 0; i < strlen(data); i++)
			if (data[i] < '0' || data[i] > '9')
				break;
		if (data[i] == 0)
			port = data;
		else
			host = data;
	}

	if (host && strcmp(host, "*") == 0)
		host = "0.0.0.0";

	*hostp = host;
	*portp = port;
}

int resolve(char *hostport, char *dflport, int stream, int passive,
		struct addrinfo **ai, int *nai, int naimax)
{
	struct addrinfo	hints, *res;
	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = stream ? SOCK_STREAM : SOCK_DGRAM;
	hints.ai_protocol = stream ? IPPROTO_TCP : IPPROTO_UDP;
	hints.ai_flags = passive ? AI_PASSIVE|AI_NUMERICHOST : AI_ADDRCONFIG;

	if (nai && *nai >= naimax)
		return -1;

	char buf[256], *host, *port;
	parse_host_port(hostport, passive, &host, &port, buf, sizeof(buf));
	if (host == NULL) {
		if (passive)
			host = "0.0.0.0";
		else
			perrordie("%s: host part missing", hostport);
	}
	if (port == NULL)
		port = dflport;

	int r = getaddrinfo(host, port, &hints, &res);
	if (r != 0)
		perrordie("%s: %s", hostport, gai_strerror(r));

	if (nai) {
		ai[*nai] = res;
		(*nai)++;
	} else {
		*ai = res;
	}

	return 0;
}

int sock_listen(char *hostport, int stream, int *sock, int *nsock, int nsmax)
{
	struct addrinfo	*res, *ai;
	resolve(hostport, "119", stream, 1, &res, NULL, 0);

	for (ai = res; ai; ai = ai->ai_next) {
		if (*nsock >= nsmax)
			break;
		int s = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
		if (s < 0)
			perrordie("socket: %m");
		int on = 1;
		if (ai->ai_family == AF_INET6) {
			setsockopt(s, IPPROTO_IPV6,
				   IPV6_V6ONLY, &on, sizeof(on));
		}
		setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
		if (bind(s, ai->ai_addr, ai->ai_addrlen) < 0)
			perrordie("socket: bind(%s): %m", hostport);
		sock[(*nsock)++] = s;
	}
	return 0;
}

void tcp_listen(char *opt, char **name, int *sock, int *nsock, int max)
{
	char buf[256];
	strncpy(buf, opt, sizeof(buf));
	buf[sizeof(buf)-1] = 0;

	char *pp;
	char *h = strtok_r(buf, ",", &pp);
	do {
		int i, n = *nsock;
		sock_listen(h, 1, sock, nsock, max);
		for (i = n; i < *nsock; i++)
			name[i] = strdup(h);
	} while ((h = strtok_r(NULL, ",", &pp)) != NULL);
}

void do_exec(int sock, char *cmd, char *remote)
{
	char host[128];
	snprintf(host, sizeof(host), "%s", "unknown");

	struct sockaddr_storage ss;
	struct sockaddr *sa = (struct sockaddr *)&ss;
	socklen_t slen = sizeof(ss);
	if (getpeername(sock, sa, &slen) == 0) {
		getnameinfo(sa, slen, host, sizeof(host), NULL, 0, 0);
	}

	char *arg0;
	if ((arg0 = strrchr(cmd, '/')) != NULL) {
		arg0++;
	} else {
		arg0 = cmd;
	}

	setenv("REALSERVERS", remote, 1);

	dup2(sock, 0);
	dup2(sock, 1);
	dup2(sock, 2);
	if (sock > 2) close(sock);
	execl(cmd, arg0, host, NULL);

	perrordie("execl(%s): %m", cmd);
}

void usage(void)
{
	fprintf(stderr, "Usage: %s -l listenaddr,[addr,...] "
			"-r remoteaddr,[remoteaddr,...] [options]\n", progname);
	fprintf(stderr, "  listenaddr:        host:port or host or port\n");
	fprintf(stderr, "  remoteaddr:        host or host:port\n");
	fprintf(stderr, "  options:\n");
	fprintf(stderr, "    -f:              foreground\n");
	fprintf(stderr, "    -p file:         pidfile\n");
	fprintf(stderr, "    -s file:         server to run (xs-nntp-slb-go)\n");
	exit(1);
}

int main(int argc, char **argv)
{
	struct pollfd		pfd[MAXLISTEN + 1];
	pid_t			pid;
	char			*pidfile = NULL;
	char			*lname[MAXLISTEN];
	int			lsock[MAXLISTEN];
	int			numlisten = 0;
	int			i, n;
	int			c;
	int			devnull;
	int			do_foreground = 0;
	char			*remote = NULL;
	char			*slb;
	char			*s;
	char			tmp[128];
	char			slb_[128];

	snprintf(tmp, sizeof(tmp), "%s", argv[0]);
	if ((s = strrchr(tmp, '/')) != NULL) {
		*s = 0;
		snprintf(slb_, sizeof(slb_), "%s/%s", tmp, "xs-nntp-slb-go");
		slb = slb_;
	}

	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];
	openlog(progname, LOG_PID, LOG_NEWS);

	for (i = 0; i < MAXLISTEN; i++)
		lsock[i] = -1;

	while ((c = getopt(argc, argv, "Ndfl:np:r:s:")) != -1) switch(c) {
		case 'N':
		case 'n':
			break;
		case 'd':
			do_debug = 1;
			do_syslog = 0;
			do_foreground = 1;
			break;
		case 'f':
			do_foreground = 1;
			break;
		case 'l':
			tcp_listen(optarg, lname, lsock, &numlisten,
					MAXLISTEN - numlisten);
			break;
		case 'p':
			pidfile = optarg;
			break;
		case 'r':
			remote = optarg;
			break;
		case 's':
			slb = optarg;
			break;
		default:
			usage();
			break;
	}

	do {
		devnull = open("/dev/null", O_RDWR);
	} while (devnull >= 0 && devnull < 3);
	if (devnull < 0)
		perrordie("/dev/null: %m");

	if (argc != optind || numlisten == 0 || remote == NULL) {
		usage();
	}

	if (!do_foreground) {
		do_stderr = 0;
		pid = fork();
		if (pid < 0)
			perrordie("fork: %m");
		if (pid > 0)
			exit(0);
		dup2(devnull, 0);
		dup2(devnull, 1);
		dup2(devnull, 2);
		setsid();
	}
	close(devnull);
	if (pidfile) {
		FILE *fp = fopen(pidfile, "w");
		if (fp) {
			fprintf(fp, "%lu\n", (unsigned long)getpid());
			fclose(fp);
		}
	}

	for (i = 0; i < numlisten; i++) {
		if (listen(lsock[i], 16) < 0)
			perrordie("listen(%s): %m", lname[i]);
		if (blocking(lsock[i], 0) < 0)
			perrordie("fcntl(O_NONBLOCK): %m", lname[i]);
		pfd[i].fd = lsock[i];
		pfd[i].events = POLLIN;
	}

	signal(SIGPIPE, SIG_IGN);
	signal(SIGCHLD, SIG_IGN);

	while (1) {
		int r = poll(pfd, numlisten, -1);
		if (r < 0) {
			if (errno == EINTR)
				continue;
			perrordie("poll: %m");
		}
		for (i = 0; i < numlisten; i++) {
			if (!(pfd[i].revents & POLLIN))
				continue;
			n = accept(pfd[i].fd, NULL, NULL);
			if (n < 0) {
				if (errno == EAGAIN)
					continue;
				if (errno == EINTR)
					continue;
				perrordie("accept: %m");
			}
			blocking(n, 1);
			pid = fork();
			if (pid == 0) {
				for (i = 0; i < numlisten; i++)
					close(lsock[i]);
				signal(SIGPIPE, SIG_DFL);
				signal(SIGCHLD, SIG_DFL);
				do_exec(n, slb, remote);
				exit(1);
			}
			if (pid < 0)
				notice("fork: %m");
			close(n);
		}
	}

	return 0;
}



main.go
package main

import (
	"crypto/md5"
	"flag"
	"fmt"
	"io"
	"net"
	"os"
	"runtime"
	"runtime/pprof"
	"strconv"
	"strings"
	"sync/atomic"
	"time"
)

const banner string = "200 %s xs-nntp-slb-go ready (transit mode)\r\n";

type NNTPCmd struct {
	name      string
	minargs   int
	maxargs   int
	fun       func(sess *NNTPSession, line string, argv []string) (error)
	help      string
}
var nntpcmds []*NNTPCmd

type NNTPStats struct {
	accepted	uint64
	refused		uint64
	rejected	uint64
	tempfail	uint64
	takethis	uint64
	ihave		uint64
}
var nntpstats NNTPStats

var nntpclients []*NNTPSession
var nntpserver *NNTPSession
var ihave_sess *NNTPSession
var startTime time.Time = time.Now()

//
//	Fast jenkins hash
//
func jenkinshash(key string) (res uint32) {
    for _, e := range key {
        res += uint32(e)
        res += (res << 10);
        res ^= (res >> 6);
    }
    res += (res << 3);
    res ^= (res >> 11);
    res += (res << 15);
    return
}

//
//	Slower MD5 hash (still pretty fast)
//	Used for compatibility with the C version.
//
func md5hash(key string) (res uint64) {
        hash := md5.Sum([]byte(key))
        res = uint64(hash[0]) + (uint64(hash[1]) << 8) +
                (uint64(hash[2]) << 16) + (uint64(hash[3]) << 24) +
                (uint64(hash[4]) << 32) + (uint64(hash[5]) << 40) +
                (uint64(hash[6]) << 48) + (uint64(hash[7]) << 56)
        return
}

func map_client(msgid string) *NNTPSession {
	return nntpclients[int(md5hash(msgid) % uint64(len(nntpclients)))]
}

func addPort(addr string, port string) (ret string) {
	ret = addr
	colons := strings.Count(addr, ":")
	if addr[0] == '[' || colons == 1 {
		return
	}
	if colons > 1 {
		ret = "[" + addr + "]:" + port
	} else {
		ret = addr + ":" + port
	}
	return
}

func updateStats(code int) {
	var ptr1, ptr2 *uint64
	switch code {
		// IHAVE
		case 235:
			ptr1 = &nntpstats.accepted
			ptr2 = &nntpstats.ihave
		case 435:
			ptr1 = &nntpstats.refused
			ptr2 = &nntpstats.ihave
		case 436:
			ptr1 = &nntpstats.tempfail
			ptr2 = &nntpstats.ihave
		case 437:
			ptr1 = &nntpstats.rejected
			ptr2 = &nntpstats.ihave
		// CHECK + TAKETHIS
		case 239:
			ptr1 = &nntpstats.accepted
			ptr2 = &nntpstats.takethis
		case 431:
			ptr1 = &nntpstats.tempfail
			ptr2 = &nntpstats.takethis
		case 438:
			ptr1 = &nntpstats.refused
			ptr2 = &nntpstats.takethis
		case 439:
			ptr1 = &nntpstats.rejected
			ptr2 = &nntpstats.takethis
		default:
	}
	if ptr1 != nil {
		atomic.AddUint64(ptr1, 1)
	}
	if ptr2 != nil {
		atomic.AddUint64(ptr2, 1)
	}
}

func logStats() {
	secs := int(time.Since(startTime).Seconds())
	n := &nntpstats
	Log.Notice("%s: stats: accepted=%d refused=%d rejected=%d " +
		"tempfail=%d takethis=%d ihave=%d seconds=%d",
		nntpserver.name,
		n.accepted, n.refused, n.rejected,
		n.tempfail, n.takethis, n.ihave, secs)
}

//
// Connect to the backend server, wait for banner, send XCLIENT,
// and expect a 200 status code.
//
func NewNNTPClient(num int, rem string) (sess *NNTPSession, err error) {
	name := fmt.Sprintf("%s:%d", rem, num)
	tmout := time.Duration(10 * time.Second)

	Log.Info("%s: connecting", name)
	conn, err := net.DialTimeout("tcp", rem, tmout)
	if err != nil {
		err = fmt.Errorf("connect: %s", err.Error())
		return
	}
	sess = NewNNTPSession(conn, name)

	conn.SetDeadline(time.Now().Add(tmout))
	line, err := sess.ReadLine()
	if err != nil {
		return
	}
	if (len(line) == 0 || line[0] != '2') {
		err = fmt.Errorf("connect failed: %s", ChompString(line))
		return
	}

	conn.SetDeadline(time.Now().Add(tmout))
	xclient := nntpserver.conn.RemoteAddr().(*net.TCPAddr).IP.String()
	err = sess.WriteAndFlush(fmt.Sprintf("XCLIENT %s\r\n", xclient))
	if err != nil {
		err = fmt.Errorf("lost connection: %s", err)
		return
	}

	conn.SetDeadline(time.Now().Add(tmout))
	line, err = sess.ReadLine()
	if err != nil {
		return
	}
	if (len(line) == 0 || line[0] != '2') {
		err = fmt.Errorf("XCLIENT failed: %s", ChompString(line))
		return
	}

	conn.SetDeadline(time.Time{})
	return
}

//
//	Send a simple reply
//
func sendreply(sess *NNTPSession, cmd string, line string) (err error) {
	req := &NNTPReq{
		line : line,
		cmd : cmd,
		ready : true,
	}
	sess.q.Add(req, true)
	return
}

//
//	Send a simple command to a backend.
//
func cmd_forward(sess *NNTPSession, c *NNTPSession, line string, arg []string, multi bool) (err error) {

	req := &NNTPReq{
		line : line,
		cmd: arg[0],
	}
	if len(arg) > 1 && arg[1][0] == '<' {
		req.msgid = arg[1]
	}

	// need to limit the amount of outstanding requests....
	// but this is pretty yucky FIXME
	for c.q.Len() > 50 {
		t := time.Duration(time.Millisecond * 10)
		time.Sleep(t)
	}

	// Add request to the main queue
	sess.q.Add(req, false)

	// Add request to the backend-specific queue
	c.q.Add(req, false)

	// And write request to backend
	if multi {
		err = c.Write(line)
		if err == nil {
			err = sess.CopyDotCRLF(c)
		}
	} else {
		err = c.WriteAndFlush(line)
	}
	return
}

func cmd_ihave(sess *NNTPSession, line string, arg []string) (err error) {
	if sess.q.Len() > 0 {
		sendreply(sess, arg[0],
			"436 This command MUST NOT be pipelined\r\n")
		return
	}
	c := map_client(arg[1])
	err = cmd_forward(sess, c, line, arg, false)
	if err == nil {
		ihave_sess = c
	}
	return
}

//
//	Send a simple command to a backend.
//
func cmd_simple(sess *NNTPSession, line string, arg []string) (err error) {
	c := map_client(arg[1])
	err = cmd_forward(sess, c, line, arg, false)
	return
}

//
//	Send a command + body to a backend
//
func cmd_withbody(sess *NNTPSession, line string, arg []string) (err error) {
	c := map_client(arg[1])
	err = cmd_forward(sess, c, line, arg, true)
	return
}

//
//	Quit command
//
func cmd_quit(sess *NNTPSession, line string, arg []string) (err error) {
	for _, c := range nntpclients {
		cmd_forward(sess, c, line, arg, false)
	}
	// note: QUIT in capitals means it won't get matched in nntpqueue.go
	if len(arg) == 1 {
		err = sendreply(sess, "QUIT", "205 Goodbye\r\n")
	}
	return
}

//
//	Help command
//
func cmd_help(sess *NNTPSession, line string, arg []string) (err error) {
	r := "100 Legal commands\r\n";
	for _, c := range nntpcmds {
		var spc string
		if (c.help == "") {
			spc = ""
		} else {
			spc = " "
		}
		r += fmt.Sprintf("  %s%s%s\r\n", c.name, spc, c.help)
	}
	r += ".\r\n"
	err = sendreply(sess, arg[0], r)
	return 
}

//
//	Capa command
//
func cmd_capa(sess *NNTPSession, line string, arg []string) (err error) {
	r := "101 Capability list:\r\n"
	r += "version 2\r\n"
	r += "implementation xs-nntp-slb-go\r\n"
	r += "ihave\r\n"
	r += "streaming\r\n"
	r += ".\r\n"
	err = sendreply(sess, arg[0], r)
	return
}

//
//	Mode command
//
func cmd_mode(sess *NNTPSession, line string, arg []string) (err error) {
	var r string
	what := strings.ToLower(arg[1])
	if (what != "stream") {
		r = "501 Unknown MODE variant\r\n"
	} else {
		r = "203 Streaming permitted\r\n"
	}
	err = sendreply(sess, arg[0], r)
	return
}

// Command dispatch table
var def_nntpcmds = []*NNTPCmd{
	&NNTPCmd{"help", 0, 0, cmd_help, ""},
	&NNTPCmd{"capabilities", 0, 1, cmd_capa, "[keyword]"},
	&NNTPCmd{"mode", 1, 1, cmd_mode, "stream"},
	&NNTPCmd{"quit", 0, 0, cmd_quit, ""},
	&NNTPCmd{"check", 1, 1, cmd_simple, "message-id"},
	&NNTPCmd{"ihave", 1, 1, cmd_ihave, "message-id"},
	&NNTPCmd{"stat", 1, 1, cmd_simple, "message-id"},
	&NNTPCmd{"takethis", 1, 1, cmd_withbody, "message-id"},
}

//
// NNTP Client: read responses from backend and queue them to be
// sent back to the remote client.
//
func run_nntpclient(sess *NNTPSession) {
	defer sess.Close()

	for {
		line, err := sess.ReadLine()
		if err != nil {
			logStats()
			Log.Fatal("%s: unexpected: %s (FATAL)", sess.name, err)
		}
		var code int64
		if len(line) > 2 {
			code, _ = strconv.ParseInt(line[0:3], 10, 16)
		}
		if code == 0 {
			Log.Error("%s: cannot parse reply code: %s",
				sess.name, line)
		}

		// So we have a reply. It corresponds to the oldest
		// command in our local queue. Pop it from the local
		// queue, and update it.
		r := sess.q.PopFirst()
		if r == nil {
			logStats()
			Log.Fatal("%s: got unexpected reply (command " +
				  "queue empty) (FATAL)", sess.name)
		}
		r.code = int(code)
		r.line = line

		if r.code > 0 {
			updateStats(r.code)
		}

		// set to ready in the global queue
		nntpserver.q.Ready(r)

		// might be a reply to the "quit" command,
		// in that case, we're done!
		if r.cmd == "quit" {
			break
		}
	}
}

func run_nntpserver(sess *NNTPSession) {

	Log.Notice("%s: connected", sess.name)

	hostname, _ := os.Hostname()
	err := sess.WriteAndFlush(fmt.Sprintf(banner, hostname))
	if err != nil {
		Log.Error("%s: unexpected: %s", sess.name, err.Error())
		return
	}

	for {
		line, err := sess.ReadLine()
		if err != nil {
			if err == io.EOF && sess.q.Len() == 0 {
				Log.Notice("%s: EOF", sess.name)
				// send QUIT to backends.
				cmd_quit(sess, "quit\r\n",
					[]string{"quit", "quiet"})
				break
			}
			logStats()
			Log.Fatal("%s: unexpected: %s (qlen=%d) (FATAL)",
				sess.name, err.Error(), sess.q.Len())
		}
		lastcode := sess.q.LastCode()
		if ihave_sess != nil && lastcode == 335 {
			//
			// Last code we saw was a 335 reply
			// to IHAVE - forward article now.
			//
			arg := []string{ "ihave" }
			err = cmd_forward(sess, ihave_sess, line, arg, true)
			if err != nil {
				logStats()
				Log.Fatal("%s: error during IHAVE forward" +
					  " to %s: %s (FATAL)", sess.name,
					  ihave_sess.name, err.Error())
			}
			ihave_sess = nil
			continue
		}
		ihave_sess = nil

		words := strings.Fields(line)
		if len(words) == 0 {
			// most NNTP servers seem to ignore empty lines
			continue
		}

		words[0] = strings.ToLower(words[0])
		cmd := words[0]
		nargs := len(words) - 1
		found := false
		for _, c := range nntpcmds {
			if (c.name != cmd) {
				continue
			}
			found = true
			if (nargs < c.minargs || nargs > c.maxargs) {
				Log.Error("%s: syntax error: %s", sess.name, line)
				sendreply(sess, cmd, "435 syntax error\r\n")
			} else {
				err = c.fun(sess, line, words)
				if err != nil {
					logStats()
					Log.Fatal("%s: error on %s: %s (FATAL)",
					sess.name, words[0], err.Error())
				}
			}
		}
		if (found == false) {
			Log.Error("%s: unknown command: %s", sess.name, line)
			sendreply(sess, cmd, "500 What?\r\n")
		}
		if cmd == "quit" {
			Log.Notice("%s: QUIT", sess.name)
			break
		}
	}
	logStats()
}

func main() {
	nntpcmds = def_nntpcmds

	var gomaxprocs int 
	var cpuprofile string
	var remote string
	var listen string

	Log.SetOutput(LogSyslog|LogStderr)

	flag.IntVar(&gomaxprocs, "gomaxprocs", 1, "number of threads")
	flag.StringVar(&cpuprofile, "cpuprofile", "", "filename.prof")
	flag.StringVar(&remote, "backend", "", "ip:port[,ip:port...]")
	flag.StringVar(&listen, "listen", "", "ip:port")
	flag.Parse()

	runtime.GOMAXPROCS(gomaxprocs)

	if cpuprofile != "" {
		f, err := os.Create(cpuprofile)
		if err != nil {
			Log.Fatal(err.Error())
		}
		pprof.StartCPUProfile(f)
		defer f.Close()
		defer pprof.StopCPUProfile()
	}

	var conn net.Conn
	var err error
	if len(listen) > 0 {
		Log.SetOutput(LogStderr)
		a, err := net.ResolveTCPAddr("tcp", listen)
		if err != nil {
			Log.Fatal(err.Error())
		}
		l, err := net.ListenTCP("tcp", a)
		if err != nil {
			Log.Fatal(err.Error())
		}
		conn, err = l.Accept()
	} else {
		Log.SetOutput(LogSyslog)
		conn, err = net.FileConn(os.Stdin)
	}
	if err != nil {
		Log.Fatal(err.Error())
	}

	rem := conn.RemoteAddr().(*net.TCPAddr).IP.String()
	names, err := net.LookupAddr(rem)
	if err == nil && len(names) > 0 && len(names[0]) > 1 {
		rem = names[0]
		l := len(rem) - 1
		if rem[l] == '.' {
			rem = rem[:l]
		}
	}
		
	nntpserver = NewNNTPSession(conn, rem)
	nntpserver.q.sess = nntpserver

	// connect to all remote servers
	if len(remote) == 0 {
		remote = os.Getenv("REALSERVERS")
	}
	if len(remote) == 0 {
		Log.Fatal("-backend and $REALSERVERS not set")
	}
	var num int
	for _, rem := range strings.Split(remote, ",") {
		num++
		rem = addPort(rem, "119")
		s, err := NewNNTPClient(num, rem)
		if err != nil {
			for _, c := range nntpclients {
				c.Close()
			}
			nntpserver.CloseMsg("500 backend " + rem + ": " +
						err.Error() + "\r\n")
			Log.Fatal("%s:%d: %s (FATAL)", rem, num, err.Error())
		}
		nntpclients = append(nntpclients, s)
	}

	doneChan := make(chan bool)
	for _, c := range nntpclients {
		go func (c *NNTPSession) {
			run_nntpclient(c)
			doneChan <- true
		}(c)
	}

	run_nntpserver(nntpserver)

	// Wait for all backends to QUIT
	Log.Info("%s: waiting for backends to shut down", nntpserver.name)

	var timeout bool
	timeChan := time.NewTimer(time.Second * 10).C

	for n := 0; n < len(nntpclients); n++ {
		select {
			case <- doneChan:
				// nothing, just loop
			case <- timeChan:
				timeout = true
				break
		}
	}

	if timeout {
		Log.Error("%s: timeout waiting for backend(s) to close",
				nntpserver.name)
	} else {
		nntpserver.q.Run()
		nntpserver.Close()
	}

	Log.Notice("%s: exit", nntpserver.name)
}

